### DISCUSSION TITLE
[20 Marzo 2023] - Pipeline

### DISCUSSION THREAD

**Discussion by notedo** [CS-Swap/Architettura-degli-Elaboratori#22] [Pipeline]
### Traccia Esercizio

Il programma qui di seguito effettua la somma dei valori (in $s0) degli elementi presenti nell’array A di lunghezza L, dopo averne modificato i valori negativi moltiplicandoli per 2. Se il risultato è maggiore o uguale a 10, viene stampata la stringa S.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
1) le istruzioni tra le quali sono presenti data hazard – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
2) le istruzioni tra le quali sono presenti control hazard – indicare i numeri di istruzione N ed M (nel formato N / M);
3) quanti cicli di clock sono necessari ad eseguire il programma tramite forwarding, spiegando il calcolo effettuato;
4) quanti cicli di clock sarebbero necessari ad eseguire il programma senza forwarding, spiegando il calcolo effettuato;
5) quali sono, per ognuna delle cinque fasi, le istruzioni (o le bolle) in pipeline durante il 13° ciclo di clock (con forwarding);
6) in quale fase la decisione delle istruzioni di jump sia compiuta nel diagramma qui sotto, e quante fasi di stallo causi potenzialmente ogni control hazard.
<img width="390" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/d16576dc-2562-4c55-b8ae-178e4b0f8f13">
<img width="870" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/1efa803c-2b69-4886-9b03-a64894292418">

### Soluzione Prof:  
[AE.2021-22.-.Appello.S2.-.2023-03-20.-.Esercizio.pipeline.-.Scheda.di.calcolo.pdf](https://github.com/user-attachments/files/15518273/AE.2021-22.-.Appello.S2.-.2023-03-20.-.Esercizio.pipeline.-.Scheda.di.calcolo.pdf)

**Comment by thegeek-sys** [comment]
# Senza forwarding
<img width="808" alt="Screenshot 2024-06-03 alle 23 20 13" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/f11c87ce-a1eb-438a-a8de-a6c33e93705a">

# Con forwarding
<img width="774" alt="Screenshot 2024-06-03 alle 22 01 06" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/4142d380-1c7e-4d47-9740-cda0480efd55">

# Data hazard e control hazard
<img width="254" alt="Screenshot 2024-06-03 alle 22 01 46" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/db04d1db-057f-4fc6-9626-7c56c49829f2">

# Istruzioni 13° ciclo di clock
<img width="92" alt="Screenshot 2024-06-03 alle 22 02 39" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/285ae425-556c-438f-a6a0-409117a104d1">
